ansible_events_ui/db/utils/lostream.py

```diff
@@ -0,0 +1,227 @@
+from __future__ import annotations
+
+import logging
+from io import UnsupportedOperation
+from typing import Tuple, Union
+
+from fastapi import Depends
+from sqlalchemy.ext.asyncio import AsyncSession
+
+from ansible_events_ui.db.dependency import get_db_session
+
+LOG = logging.getLogger(__name__)
+
+
+# Default read buffer size is
+CHUNK_SIZE = 1024**2
+
+
+class LObject:
+    """
+    Facilitate PostgreSQL large object interface using server-side functions.
+
+    As of 2022-09-01, there is no large object support directly
+    in asyncpg or other async Python PostgreSQL drivers.
+    """
+
+    VALID_MODES = {"rwt", "rt", "wt", "at", "rwb", "rb", "wb", "ab"}
+    INV_READ = 0x00040000
+    INV_WRITE = 0x00020000
+    MODE_MAP = {"r": INV_READ, "w": INV_WRITE, "a": INV_READ | INV_WRITE}
+
+    def __init__(
+        self,
+        oid: int = 0,
+        length: int = 0,
+        session: AsyncSession = None,
+        *,
+        chunk_size: int = CHUNK_SIZE,
+        mode: str = "rb",
+    ):
+        self.session = session
+        self.chunk_size = chunk_size if chunk_size > 0 else CHUNK_SIZE
+        self.oid = oid
+        self.length = length
+        self.closed = False
+
+        self.imode, self.text_data, self.append = self._resolve_mode(mode)
+        self.pos = self.length if self.append else 0
+
+    def _resolve_mode(self, mode: str):
+        if mode not in self.VALID_MODES:
+            raise ValueError(
+                f"Mode {mode} must be one of {sorted(self.VALID_MODES)}"
+            )
+
+        _text_data = mode.endswith("t")
+        _append = mode.startswith("a")
+        _mode = 0
+        for c in mode:
+            _mode |= self.MODE_MAP.get(c, 0)
+
+        return _mode, _text_data, _append
+
+    def closed_check(self):
+        if self.closed:
+            raise UnsupportedOperation("Large object is closed")
```

Comment:
If the intention of this class was to replicate behavior of a standard python IO stream, the operation on a closed file should raise `ValueError`.